Meta+LibCore: Include git hash in version string

tcl3 opened the pull request:
The `--version` argument will now append the git hash to the version number if it is known.

Currently, a rebuild is required to reliably update the generated git hash.

trflynn89 commented:
I think we've talked about generating git hashes like this before and have been actively avoiding it. For one, this only runs when CMake itself runs - if we just make a commit to a LibWeb .cpp file, that won't actually cause GitVersion.h to be regenerated:
```
» git --no-pager log --oneline -n 3
2bb8ecd970 (HEAD -> font_face_set_load) empty commit
cc903872be LibWeb: Begin implementing FontFaceSet.prototype.load
00c45243bd (upstream/master, master) LibRegex: Don't blindly accept inverted charclasses for atomic rewrite

» br --version
ninja: Entering directory `/Users/flynn/workspace/ladybird/Build/ladybird'
[0/2] Re-checking globbed directories...
[9/10] Linking CXX executable bin/Ladybird.app/Contents/MacOS/Ladybird

Version 1.0-cc903872be
```

Can whatever script that is used to run WPT on the dedicated runner do a `git log` command to get the hash instead?

tcl3 commented:
> #858 my related previous attempt (and record of discussion around it)

I completely forgot that this had already been attempted. Most of the discussion there also applies to this attempt

> Can whatever script that is used to run WPT on the dedicated runner do a `git log` command to get the hash instead?

wpt.fyi uses the `browser_version` field of the wptrport json file to show the browser version. This field is populated by this code:

https://github.com/web-platform-tests/wpt/blob/2a6f629b52f29e5431b1937d067c81bf5db73d39/tools/wpt/browser.py#L2254-L2264

It's also possible to specify the `browser_version` when uploading to wpt.fyi:  https://github.com/web-platform-tests/wpt.fyi/blob/main/api/README.md#results-creation, but apparently this only works if the `browser_version` field hasn't already been populated in the json file.

We could modify the wptreport.json file before uploading it to wpt.fyi, or update/remove the version-getting code above, but IMO, these options are worse than having some kind of mechanism to make `--version` show the git hash.

trflynn89 commented:
I was curious what Servo does. Turns out they do embed the git hash in the `--version` string, but they get it via an environment variable:
https://github.com/servo/servo/blob/8e2a0414e870ffbca620a05ea9a18878e81e6622/ports/servoshell/lib.rs#L79

I don't see any other reference to that var in any of their repos, other than that they initialize it to a "no git" string in their build environment. But presumably they are setting it from their WPT runner.

While that is certainly a weird approach, maybe we can do something similar for now? (I would also add sanity checks after getting the variable, like it has the expected length / is an alphanum string).

sideshowbarker commented:
> I was curious what Servo does. Turns out they do embed the git hash in the `--version` string, but they get it via an environment variable:

FWIW, my initial commit for #858 at https://github.com/LadybirdBrowser/ladybird/commit/6346ed08d1f02a546004d6cfb803012ee0b7cc55 worked by setting an environment variable in the build environment, and then using that later.

I rewrote it because the use case I had for my PR wasn’t limited to only exposing the version data just in the build environment. But if the use case here _is_ limited to that, then on the face of it, using an environment variable seems like a reasonable way to do it.

tcl3 commented:
I've updated the PR to use the `GIT_VERSION` environment variable for this. I've also updated `WPT.sh` to ensure the environment variable is set when invoking `wpt`.

tcl3 commented on Userland/Libraries/LibCore/Version.cpp:
```diff
@@ -5,12 +5,16 @@
 {
+    auto maybe_git_version = Core::Environment::get("GIT_VERSION"sv);
+    if (maybe_git_version.has_value())
+        return MUST(String::formatted("Version 1.0-{}", maybe_git_version.value()));
```
I've updated the environment variable name to `LADYBIRD_GIT_VERSION`  and added validation to ensure that only valid git hashes will be printed. The validation function checks for lengths between 4 and 40 chars inclusive, as `git rev-parse --short=N` only guarantees a minimum of N characters will be used (with an absolute minimum of 4 characters).
